pkg/controller/atlasproject/teams.go

```diff
@@ -30,8 +31,56 @@ type TeamDataContainer struct {
 	Context     *workflow.Context
 }
 
+func (r *AtlasProjectReconciler) garbageCollectTeams(ctx context.Context, projectID string) error {
+	teams := &akov2.AtlasTeamList{}
+	err := r.Client.List(ctx, teams)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve a list of teams: %w", err)
+	}
+	if len(teams.Items) == 0 {
+		return nil
+	}
+	g, ctx := errgroup.WithContext(ctx)
+	for _, t := range teams.Items {
+		team := t
+		g.Go(func() error {
+			for _, project := range team.Status.Projects {
+				if project.ID != projectID {
+					continue
+				}
+
+				team.UpdateStatus([]status.Condition{}, status.AtlasTeamUnsetProject(projectID))
+
+				if err = r.Client.Status().Update(ctx, &team); err != nil {
+					r.Log.Errorw("failed to update Team status", "error", err)
+					return err
+				}
+
+				if len(team.Status.Projects) == 0 &&
+					customresource.HaveFinalizer(&team, customresource.FinalizerLabel) {
+					customresource.UnsetFinalizer(&team, customresource.FinalizerLabel)
+				}
+
+				if err = r.Client.Update(ctx, &team); err != nil {
+					r.Log.Errorw("failed to update Team object", "error", err)
+					return err
+				}
+			}
+			return nil
+		})
+	}
+
+	return g.Wait()
+}
+
 func (r *AtlasProjectReconciler) ensureAssignedTeams(workflowCtx *workflow.Context, project *akov2.AtlasProject, protected bool) workflow.Result {
 	resourcesToWatch := make([]watch.WatchedObject, 0, len(project.Spec.Teams))
+
+	err := r.garbageCollectTeams(workflowCtx.Context, project.ID())
+	if err != nil {
+		return workflow.Terminate(workflow.TeamNotCleaned, err.Error())
+	}
```

Comment:
Calling the garbage collector directly here will remove the project from team status even if it wasn't de-assigned.
While the next part of the code will add it back, this may have a workflow that leads to a new bug.
Here I'd call the garage collector only when `project.Spec.Teams` is empty
```suggestion
        if len(project.Spec.Teams) == 0 {
 	    err := r.garbageCollectTeams(workflowCtx.Context, project.ID())
	    if err != nil {
	 	    return workflow.Terminate(workflow.TeamNotCleaned, err.Error())
	    }
	}
```

Reply:
What if only one team was removed? E.g. there are two teams, one got removed. In this case, it won't be garbage collected.

Reply:
I see... with that, we need to be sure a cleaned-up team gets its status corrected later or do a separate clean-up function for this case. If the current implementation already covers first mentioned option, I'm good

Reply:
Let me add one more test to make sure we cover this scenario as well